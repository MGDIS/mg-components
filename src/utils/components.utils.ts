/**
 * Create random ID
 *
 * @param {string} prefix  add prefix to created ID
 * @param {number} length ID length
 * @returns {string} ID
 */
export const createID = (prefix = '', length = 10): string => {
  let ID = '';
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  const charsLength = chars.length;
  for (let i = 0; i < length; i++) {
    ID += chars.charAt(Math.floor(Math.random() * charsLength));
  }
  return (prefix !== '' ? `${prefix}-` : '') + ID;
};

/**
 * Class to manage component classlist
 * Set() are not working when imported in project
 */
export class ClassList {
  classes: string[];

  constructor(classlist: string[] = []) {
    this.classes = classlist;
  }

  /**
   * Add class
   *
   * @param {string} className class name to add
   * @returns {void}
   */
  add = (className: string): void => {
    if (!this.has(className)) {
      this.classes.push(className);
    }
  };

  /**
   * Delete class
   *
   * @param {string} className class name to delete
   * @returns {void}
   */
  delete = (className: string): void => {
    const index = this.classes.indexOf(className);
    if (index > -1) {
      this.classes.splice(index, 1);
    }
  };

  /**
   * Check if class exist in list
   *
   * @param {string} className class name to check
   * @returns {boolean} class name is in the list
   */
  has = (className: string): boolean => {
    return this.classes.includes(className);
  };

  /**
   * Join classes seperated by spaces
   *
   * @returns {string} joined values
   */
  join = (): string => {
    return this.classes.join(' ');
  };
}

/**
 * Check if all items are string
 *
 * @param {string[]} items items to check
 * @returns {boolean} all items are string
 */
export const allItemsAreString = (items: string[]): boolean => {
  return items && items.every(item => typeof item === 'string');
};

/**
 * Check if element is a heading
 *
 * @param {Element} element slotted element
 * @param {string[]} tagNames allowed tag names list
 * @returns {boolean} element is a heading
 */
export const isTagName = (element: Element, tagNames: string[]): boolean => {
  return tagNames.includes(element?.tagName.toLowerCase());
};

/**
 * Focusable elements query selector
 */
export const focusableElements = 'a[href], button, input, textarea, select, details, [tabindex]:not([tabindex="-1"]), [identifier]';

/**
 *
 * @param {Window} localWindow the window we are lookink for other windows
 * @returns {Window[]} The list of windows found
 */
export const getWindows = (localWindow: Window): Window[] => {
  const parentWindows = getParentWindows(localWindow);
  const childWindows = getChildWindows(localWindow);
  return [localWindow, ...parentWindows, ...childWindows];
};

/**
 * Get parent windows
 *
 * @param {Window} localWindow the window we are lookink for parents
 * @param {Window[]} windows The list of allready found windows
 * @returns {Window[]} The list of windows found
 */
export const getParentWindows = (localWindow: Window, windows: Window[] = []): Window[] => {
  // Check if is in iframe
  if (localWindow.self !== localWindow.top) {
    // Check if we have permission to access parent
    try {
      const parentWindow: Window = localWindow.parent;
      windows.push(parentWindow);
      return getParentWindows(parentWindow, windows);
    } catch (err) {
      console.error('Different hosts between iframes:', err);
      return windows;
    }
  }
  return windows;
};

/**
 * Get child windows
 *
 * @param {Window} localWindow the window we are lookink for children
 * @param {Window[]} windows The list of allready found windows
 * @returns {Window[]} The list of windows found
 */
const getChildWindows = (localWindow: Window, windows: Window[] = []): Window[] => {
  if (localWindow.frames.length > 0)
    for (let i = 0; i < localWindow.frames.length; i++) {
      const childWindow: Window = localWindow.frames[i];
      windows.push(childWindow);
      return getChildWindows(childWindow, windows);
    }
  return windows;
};
