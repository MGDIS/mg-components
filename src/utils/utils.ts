/**
 * Create random ID
 * @param prefix {string} add prefix to created ID
 * @returns {string} ID
 */
export function createID(prefix?: string): string {
  return (typeof prefix === 'string' && prefix !== '' ? `${prefix}-` : '' ) + Math.random().toString(36).substr(2, 10);
}

/**
 * Change date format
 * @param date {string} date with format yyyy-mm-dd
 * @returns {string} date with format dd/mm/yyyy
 */
export function formatDate(date: string): string {
  const dateRegexp = /\d{4}-\d{2}-\d{2}/;
  if(typeof date !== 'string' || date === '' || !dateRegexp.test(date)) {
    return '';
  }
  const d:Date = new Date(date);
  return `${d.getDate()}/${d.getMonth()+1}/${d.getFullYear()}`;
}

/**
 * Class to manage component classlist
 * Set() are not working when imported in project
 */
export class ClassList {
  classes: string[];

  constructor (classlist:string[] = []) {
    this.classes = classlist;
  }

  /**
   * Add class
   * @param className
   * @returns {void}
   */
  add = (className:string) => {
    if(!this.classes.includes(className)) {
      this.classes.push(className);
    }
  }

  /**
   * Delete class
   * @param className
   * @returns {void}
   */
  delete = (className:string) => {
    const index = this.classes.indexOf(className);
    if(index > -1) {
      this.classes.splice(index, 1);
    }
  }

  /**
   * Join classes seperated by spaces
   * @returns {string}
   */
  join = ():string => {
    return this.classes.join(' ');
  }
}
