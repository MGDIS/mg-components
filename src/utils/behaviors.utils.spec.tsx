import { h } from '@stencil/core/internal';
import { newSpecPage } from '@stencil/core/testing';
import { MgMenuItem } from '../components/molecules/menu/mg-menu-item/mg-menu-item';
import { MgMenu } from '../components/molecules/menu/mg-menu/mg-menu';
import { OverflowBehavior } from './behaviors.utils';
import { setupResizeObserverMock } from './unit.test.utils';

const getPage = args =>
  newSpecPage({
    components: [MgMenu, MgMenuItem],
    template: () => (
      <mg-menu {...args}>
        <mg-menu-item>
          <span slot="label">Batman</span>
        </mg-menu-item>
        <mg-menu-item>
          <span slot="label">Joker</span>
        </mg-menu-item>
        <mg-menu-item>
          <span slot="label">Bane</span>
        </mg-menu-item>
      </mg-menu>
    ),
  });

describe('behavior.utils', () => {
  describe('OverflowBehavior', () => {
    let fireRo;
    beforeEach(() => {
      jest.useFakeTimers();
      setupResizeObserverMock({
        observe: function () {
          fireRo = this.cb;
        },
        disconnect: function () {
          return null;
        },
      });
    });
    afterEach(() => jest.runOnlyPendingTimers());

    describe('render', () => {
      test.each([50, 80, 100, 250])('should manage resize with observer, case %s', async width => {
        const page = await getPage({ label: 'batman' });
        const element = page.doc.querySelector('mg-menu');

        const items = Array.from(page.doc.querySelectorAll('mg-menu-item'));
        items.forEach(item => {
          Object.defineProperty(item, 'offsetWidth', {
            get: jest.fn(() => 80),
          });
        });

        const createMenuItem = (label: string) => {
          const menuItem = page.doc.createElement('mg-menu-item');
          const span = page.doc.createElement('span');
          span.innerText = label;
          menuItem.appendChild(span);
          return menuItem;
        };
        const createMenu = (label: string) => {
          const menu = page.doc.createElement('mg-menu');
          menu.label = label;
          return menu;
        };
        const renderMoreElement = () => {
          const moreMenu = 'more menu';
          const moreElement = createMenuItem(moreMenu);
          const moreMenuElement = createMenu(moreMenu);
          moreMenuElement.appendChild(createMenuItem('proxy batman'));
          moreMenuElement.appendChild(createMenuItem('proxy joker'));
          moreMenuElement.appendChild(createMenuItem('proxy bane'));
          moreElement.appendChild(moreMenuElement);
          Object.defineProperty(moreElement, 'offsetWidth', {
            get: jest.fn(() => 50),
          });
          element.append(moreElement);
          return moreElement;
        };

        new OverflowBehavior(element, renderMoreElement);

        fireRo([{ contentRect: { width } }]);

        expect(page.root).toMatchSnapshot();
      });
    });

    test('should fire disconnect callback', () => {
      const behavior = new OverflowBehavior(<div></div>, () => <span></span>);
      const spy = jest.spyOn((behavior as any).resizeObserver, 'disconnect');
      behavior.disconnect();

      expect(spy).toHaveBeenCalled();
    });
  });
});
