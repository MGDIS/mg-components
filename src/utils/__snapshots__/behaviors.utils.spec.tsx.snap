// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`behavior.utils OverflowBehavior render menu should manage resize with observer, case 50 1`] = `
<mg-menu aria-label="batman" direction="horizontal" role="menubar">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="0" data-style-direction-horizontal role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Batman">
      Batman
    </span>
  </mg-menu-item>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="1" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Joker">
      Joker
    </span>
  </mg-menu-item>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="2" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Bane">
      Bane
    </span>
  </mg-menu-item>
  <mg-item-more aria-haspopup="true" role="menuitem">
    <mock:shadow-root>
      <mg-menu-item data-overflow-more data-size="regular" status="visible">
        <mock:shadow-root>
          <button aria-expanded="false" class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
            <slot name="image"></slot>
            <div class="mg-menu-item__navigation-button-center">
              <div class="mg-menu-item__navigation-button-text-content">
                <slot name="label"></slot>
                <slot name="information"></slot>
              </div>
            </div>
            <span class="mg-a11y-animation mg-menu-item__navigation-button-chevron">
              <mg-icon icon="chevron-down"></mg-icon>
            </span>
          </button>
          <div class="mg-menu-item__collapse-container" hidden="">
            <slot></slot>
          </div>
        </mock:shadow-root>
        <mg-icon icon="ellipsis-vertical" slot="image"></mg-icon>
        <span class="sr-only" slot="label" title="Additional menu">
          Additional menu
        </span>
        <mg-menu aria-label="Additional menu" direction="vertical" role="menubar">
          <mock:shadow-root>
            <slot></slot>
          </mock:shadow-root>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="0" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Batman">
              Batman
            </span>
          </mg-menu-item>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="1" data-style-direction-horizontal role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Joker">
              Joker
            </span>
          </mg-menu-item>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="2" data-style-direction-horizontal role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Bane">
              Bane
            </span>
          </mg-menu-item>
        </mg-menu>
      </mg-menu-item>
    </mock:shadow-root>
  </mg-item-more>
</mg-menu>
`;

exports[`behavior.utils OverflowBehavior render menu should manage resize with observer, case 80 1`] = `
<mg-menu aria-label="batman" direction="horizontal" role="menubar">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="0" data-style-direction-horizontal role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Batman">
      Batman
    </span>
  </mg-menu-item>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="1" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Joker">
      Joker
    </span>
  </mg-menu-item>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="2" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Bane">
      Bane
    </span>
  </mg-menu-item>
  <mg-item-more aria-haspopup="true" role="menuitem">
    <mock:shadow-root>
      <mg-menu-item data-overflow-more data-size="regular" status="visible">
        <mock:shadow-root>
          <button aria-expanded="false" class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
            <slot name="image"></slot>
            <div class="mg-menu-item__navigation-button-center">
              <div class="mg-menu-item__navigation-button-text-content">
                <slot name="label"></slot>
                <slot name="information"></slot>
              </div>
            </div>
            <span class="mg-a11y-animation mg-menu-item__navigation-button-chevron">
              <mg-icon icon="chevron-down"></mg-icon>
            </span>
          </button>
          <div class="mg-menu-item__collapse-container" hidden="">
            <slot></slot>
          </div>
        </mock:shadow-root>
        <mg-icon icon="ellipsis-vertical" slot="image"></mg-icon>
        <span class="sr-only" slot="label" title="Additional menu">
          Additional menu
        </span>
        <mg-menu aria-label="Additional menu" direction="vertical" role="menubar">
          <mock:shadow-root>
            <slot></slot>
          </mock:shadow-root>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="0" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Batman">
              Batman
            </span>
          </mg-menu-item>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="1" data-style-direction-horizontal role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Joker">
              Joker
            </span>
          </mg-menu-item>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="2" data-style-direction-horizontal role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Bane">
              Bane
            </span>
          </mg-menu-item>
        </mg-menu>
      </mg-menu-item>
    </mock:shadow-root>
  </mg-item-more>
</mg-menu>
`;

exports[`behavior.utils OverflowBehavior render menu should manage resize with observer, case 100 1`] = `
<mg-menu aria-label="batman" direction="horizontal" role="menubar">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="0" data-style-direction-horizontal role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Batman">
      Batman
    </span>
  </mg-menu-item>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="1" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Joker">
      Joker
    </span>
  </mg-menu-item>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="2" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Bane">
      Bane
    </span>
  </mg-menu-item>
  <mg-item-more aria-haspopup="true" role="menuitem">
    <mock:shadow-root>
      <mg-menu-item data-overflow-more data-size="regular" status="visible">
        <mock:shadow-root>
          <button aria-expanded="false" class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
            <slot name="image"></slot>
            <div class="mg-menu-item__navigation-button-center">
              <div class="mg-menu-item__navigation-button-text-content">
                <slot name="label"></slot>
                <slot name="information"></slot>
              </div>
            </div>
            <span class="mg-a11y-animation mg-menu-item__navigation-button-chevron">
              <mg-icon icon="chevron-down"></mg-icon>
            </span>
          </button>
          <div class="mg-menu-item__collapse-container" hidden="">
            <slot></slot>
          </div>
        </mock:shadow-root>
        <mg-icon icon="ellipsis-vertical" slot="image"></mg-icon>
        <span class="sr-only" slot="label" title="Additional menu">
          Additional menu
        </span>
        <mg-menu aria-label="Additional menu" direction="vertical" role="menubar">
          <mock:shadow-root>
            <slot></slot>
          </mock:shadow-root>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="0" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Batman">
              Batman
            </span>
          </mg-menu-item>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="1" data-style-direction-horizontal role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Joker">
              Joker
            </span>
          </mg-menu-item>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="2" data-style-direction-horizontal role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Bane">
              Bane
            </span>
          </mg-menu-item>
        </mg-menu>
      </mg-menu-item>
    </mock:shadow-root>
  </mg-item-more>
</mg-menu>
`;

exports[`behavior.utils OverflowBehavior render menu should manage resize with observer, case 250 1`] = `
<mg-menu aria-label="batman" direction="horizontal" role="menubar">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="0" data-style-direction-horizontal role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Batman">
      Batman
    </span>
  </mg-menu-item>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="1" data-style-direction-horizontal role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Joker">
      Joker
    </span>
  </mg-menu-item>
  <mg-menu-item aria-haspopup="false" data-overflow-base-index="2" data-style-direction-horizontal role="menuitem" status="visible">
    <mock:shadow-root>
      <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--horizontal mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
        <slot name="image"></slot>
        <div class="mg-menu-item__navigation-button-center">
          <div class="mg-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-menu-item__collapse-container mg-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </mock:shadow-root>
    <span slot="label" title="Bane">
      Bane
    </span>
  </mg-menu-item>
  <mg-item-more aria-haspopup="true" hidden="" role="menuitem">
    <mock:shadow-root>
      <mg-menu-item data-overflow-more data-size="regular" status="visible">
        <mock:shadow-root>
          <button aria-expanded="false" class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
            <slot name="image"></slot>
            <div class="mg-menu-item__navigation-button-center">
              <div class="mg-menu-item__navigation-button-text-content">
                <slot name="label"></slot>
                <slot name="information"></slot>
              </div>
            </div>
            <span class="mg-a11y-animation mg-menu-item__navigation-button-chevron">
              <mg-icon icon="chevron-down"></mg-icon>
            </span>
          </button>
          <div class="mg-menu-item__collapse-container" hidden="">
            <slot></slot>
          </div>
        </mock:shadow-root>
        <mg-icon icon="ellipsis-vertical" slot="image"></mg-icon>
        <span class="sr-only" slot="label" title="Additional menu">
          Additional menu
        </span>
        <mg-menu aria-label="Additional menu" direction="vertical" role="menubar">
          <mock:shadow-root>
            <slot></slot>
          </mock:shadow-root>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="0" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Batman">
              Batman
            </span>
          </mg-menu-item>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="1" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Joker">
              Joker
            </span>
          </mg-menu-item>
          <mg-menu-item aria-haspopup="false" data-overflow-proxy-index="2" data-style-direction-horizontal hidden="" role="menuitem" status="visible">
            <mock:shadow-root>
              <button class="mg-menu-item__navigation-button mg-menu-item__navigation-button--size-regular mg-menu-item__navigation-button--visible" tabindex="0">
                <slot name="image"></slot>
                <div class="mg-menu-item__navigation-button-center">
                  <div class="mg-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </mock:shadow-root>
            <span slot="label" title="Bane">
              Bane
            </span>
          </mg-menu-item>
        </mg-menu>
      </mg-menu-item>
    </mock:shadow-root>
  </mg-item-more>
</mg-menu>
`;
